### PR TITLE
Add `--sharded` parameter to only run a subset of packages on a worker

### PR DESCRIPTION
**Pull request by weswigham**
It takes two numbers, the shard id and the total shard count. Usage is `--sharded 1 3`. The current sharding method is naive, but without historical timing data, it's pretty OK, and should work well provided most packages take roughly the same time to check.

This will allow us to parallelize DT runs over as many containers as we'd like, like we'd discussed at lunch the other day :heart:

### CODE REVIEW

**Comment by weswigham** [comment]
@sandersn Wanna give this a once over and merge it?

Ah, also, while there's nothing wrong with how this sharding is done, there _is_ a race condition in how DT is cloned:
If you spawn multiple workers and they snap different DT versions, you could get inconsistent results. If the build task is started on a DT trigger (so `--clone` isn't used) this isn't a problem, however if you then use `--localTs` and clone a TS version in your pipeline, if that version isn't set (eg, it's a branch name like `master`), then _that_ can have a race condition. Synchronizing the clones across jobs can be done at the job-orchestrator level, though, eg, by in a non-parallel phase executed before the parallel one, getting the commit SHA of each thing that needs to be cloned and then setting it for the entire pipeline. Supporting this for the `--clone` param may need a little more work, to allow passing in a DT commit sha to clone.

**Comment by sandersn** [comment]
Is the purpose of this making DT runs faster for us, or for DT contributors? This only improves it for us.

On the other hand, the median DT contributor wouldn't benefit from sharding, contra the median TS contributor who wants to run DT. Maybe we can leave types-publisher sharding until we start getting close to the 6 hour limit for react. :)

**Comment by weswigham** [comment]
For us, mostly. `@typescript-bot run dt` takes an hour and a half right now.

**Comment by weswigham** [comment]
(And in a lot of cases it'd be real nice to get that feedback for iteration faster)

File: src/index.ts [hunk 1]
```diff
@@ -155,7 +168,7 @@ const exclude = new Set<string>([
     "strophe",
 ]);
 
-async function getAllPackages(typesDir: string): Promise<ReadonlyArray<string>> {
+async function getAllPackages(typesDir: string, sharding: {shardId: number, shardCount: number} | undefined): Promise<ReadonlyArray<string>> {
```
Comment (sandersn) [inline src/index.ts]: can you de-prefix the type to `{ id: number, count: number }` ? It's obvious from useage everywhere what the names refer to.

File: src/index.ts [hunk 2]
```diff
@@ -48,12 +50,23 @@ if (module.parent === null) { // tslint:disable-line no-null-keyword
                 assert(!Number.isNaN(nProcesses));
                 break;
             }
+            case "--sharded": {
+                i++;
+                assert(i < argv.length);
+                shardId = Number.parseInt(argv[i]);
```
Comment (sandersn) [inline src/index.ts]: seems like it would be less confusing to `-1` shardId here rather than in the actual filter.

Comment (weswigham) [reply src/index.ts]: The shard is actually assigned a 1-based index by pipelines - i'd rather decrement that for the modular arithmetic operation, so as to not muddy the meaning.

File: src/index.ts (hunk 1 above)
Comment (weswigham) [reply src/index.ts]: Alright.

**Comment by weswigham** [comment]
> Have you tested this manually?

I've run commands for `--sharded 1 2` and `--sharded 2 2` and validated that they started working on different sets of packages. I did not, ofc, wait for all of those to finish validations 😄

**Comment by weswigham** [comment]
I also have a [build definition in progress](https://typescript.visualstudio.com/TypeScript/_build?definitionId=23) to make use of this. I'm currently working on preventing the possible race with `--clone`.

File: src/index.ts (hunk 2 above)
Comment (sandersn) [reply src/index.ts]: No problem; readability is subjective.

**Comment by weswigham** [comment]
@sandersn I just added the change to allow `--clone` to optionally take a ref/sha (so `--clone 0376cb`) to this PR (so it can be used to avoid the race condition I mentioned in my first comment). Wanna look over that, too?

**Comment by weswigham** [comment]
@sandersn also since this is your repo I can't actually merge this; that's up to you ❤️

File: src/index.ts [hunk 3]
```diff
@@ -155,10 +161,28 @@ async function installAllDependencies(
...
+            `git checkout ${sha}` // then check it out
+        ];
+        for (const command of commands) {
+            console.log(command);
+            await runOrFail(cwd, command);
+        }
+        return;
+    }
+    else {
+        const cmd = `git clone https://github.com/DefinitelyTyped/DefinitelyTyped.git --depth 1`;
```
Comment (sandersn) [inline src/index.ts]: what's with all the template literals???

**Comment by sandersn** [comment]
Oops, right. I forgot.

File: src/index.ts (hunk 3 above)
Comment (weswigham) [reply src/index.ts]: Backticks 4 lyfe. But mostly backtick string > double quote string when you're unsure if you're going to use interpolation or quotes in the string or not, and it's kinda strange to have a list of strings but only have one in backticks and the rest not 🤷‍♂️

**Comment by weswigham** [comment]
Alright. `@typescript-bot run dt faster` should now work. Rather than running DT on one agent, it runs it on 4. This _does_ currently have the potential for clogging up our build queue a bit more (especially during already high-load times of day), though. I also updated the bot to trigger on [regex matches](https://github.com/weswigham/typescript-bot-test-triggerer/blob/master/GithubCommentReader/index.js#L147) rather than simple `.indexOf` (to allow future tasks to read params from the comment) so if you notice any changes in behavior (like it failing to trigger on some input you expected it to trigger for), let me know~

**Comment by weswigham** [comment]
Rapid update: Browsing the output of the [trial run](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=26254) I started, it looks like transitive non-DT dependencies of packages not in the current shard aren't installed :( (So, for example, `chai-enzyme` depends on `react` (which is tested in a different shard) which in turn depends on `csstype` (a non-DT dependency, which wasn't installed because we never ran `npm i` in the `react` folder because another shard is running tests on that folder))

@sandersn is there an easy fix for this, or should I just swap the sharded mode of `dtslint-runner` to installing _all_ dependencies, rather than just dependencies for things being tested?

**Comment by sandersn** [comment]
Nope, no easy fix, as far as I know. I think you should install all dependencies.

**Pull request merged**